processor/src/trace/virtual_bus/sum_check/prover/mod.rs

```diff
@@ -0,0 +1,323 @@
+            // claim. This basically computes the s_i polynomial.
+            let round_poly_evals = sumcheck_round(&self.composition_poly, mls);
+
+            // update the claim
+            current_round_claim = new_round_claim;
+
+            let round_poly_coefs = round_poly_evals.to_poly(current_round_claim.claim);
+
+            // reseed with the s_i polynomial
+            coin.reseed(H::hash_elements(&round_poly_coefs.coefficients));
+            let round_proof = RoundProof { round_poly_coefs };
+            round_proofs.push(round_proof);
+        }
+
+        // generate the last random challenge
+        let round_challenge = coin.draw().map_err(|_| Error::FailedToGenerateChallenge)?;
+        // fold each multi-linear using the last random challenge
+        mls.iter_mut().for_each(|ml| ml.bind(round_challenge));
+
+        let round_claim =
+            reduce_claim(&round_proofs[num_rounds - 1], current_round_claim, round_challenge);
+        Ok((round_claim, round_proofs))
+    }
+}
+
+/// Computes the polynomial
+///
+/// s_i(X_i) := \sum_{(x_{i + 1},\cdots, x_{\nu - 1})
+///                                  w(r_0,\cdots, r_{i - 1}, X_i, x_{i + 1}, \cdots, x_{\nu - 1}).
+/// where
+///
+/// w(x_0,\cdots, x_{\nu - 1}) := g(f_0((x_0,\cdots, x_{\nu - 1})),
+///                                                       \cdots , f_c((x_0,\cdots, x_{\nu - 1}))).
+///
+/// Given a degree bound `d_max` for all variables, it suffices to compute the evaluations of `s_i`
+/// at `d_max + 1` points. Given that `s_{i}(0) = s_{i}(1) - s_{i - 1}(r_{i - 1})` it is sufficient
+/// to compute the evaluations on only `d_max` points.
+///
+/// The algorithm works by iterating over the variables (x_{i + 1}, \cdots, x_{\nu - 1}) in
+/// {0, 1}^{\nu - 1 - i}. For each such tuple, we store the evaluations of the (folded)
+/// multi-linears at (0, x_{i + 1}, \cdots, x_{\nu - 1}) and
+/// (1, x_{i + 1}, \cdots, x_{\nu - 1}) in two arrays, `evals_zero` and `evals_one`.
+/// Using `evals_one`, remember that we optimize evaluating at 0 away, we get the first evaluation
+/// i.e., `s_i(1)`.
+///
+/// For the remaining evaluations, we use the fact that the folded `f_i` is multi-linear and hence
+/// we can write
+///
+///     f_i(X_i, x_{i + 1}, \cdots, x_{\nu - 1}) =
+///        (1 - X_i) . f_i(0, x_{i + 1}, \cdots, x_{\nu - 1}) + X_i . f_i(1, x_{i + 1}, \cdots, x_{\nu - 1})
+///
+/// Note that we omitted writing the folding randomness for readability.
+/// Since the evaluation domain is {0, 1, ... , d_max}, we can compute the evaluations based on
+/// the previous one using only additions. This is the purpose of `deltas`, to hold the increments
+/// added to each multi-linear to compute the evaluation at the next point, and `evals_x` to hold
+/// the current evaluation at `x` in {2, ... , d_max}.
+fn sumcheck_round<E: FieldElement>(
+    composition_poly: &dyn CompositionPolynomial<E>,
+    mls: &mut [MultiLinearPoly<E>],
+) -> UnivariatePolyEvals<E> {
```

Comment:
A few questions here:

1. Why do we need dynamic objects here (i.e., `dyn`)? Could we not just use simple generics?
2. It seems like invocation of `sumcheck_round()` is always followed by `.to_poly()` call on the result. Would it make sense to move `.to_poly()` call into the `sumcheck_round()` function.
3. Why `mls` parameter needs to be passed in as `mut`? It doesn't seem like we are mutating it in the function.

Applying all 3 changes above, the signature of the function could look like so:
```Rust
fn sumcheck_round<E: FieldElement, P: CompositionPolynomial<E>>(
    composition_poly: &P,
    mls: &[MultiLinearPoly<E>],
) -> UnivariatePolyCoef<E> 
```

Reply:
1. Probably a remnant from a previous approach. Changed.
2. We can, but then we would have to pass `claim` so that we can call `to_poly`. I think keeping `sumcheck_round` clean is a good idea but, again, no strong preference from my side.
3. I doesn't. Changed

Reply:
> 2\. We can, but then we would have to pass `claim` so that we can call `to_poly`. I think keeping `sumcheck_round` clean is a good idea but, again, no strong preference from my side.

Ah - I see. Let's keep it as is for now.